library/blivet.py

```diff
@@ -0,0 +1,709 @@
+    type: list of dict
+'''
+
+import logging
+
+try:
+    from blivet3 import Blivet
+    from blivet3.callbacks import callbacks
+    from blivet3.flags import flags as blivet_flags
+    from blivet3.formats import get_format
+    from blivet3.partitioning import do_partitioning
+    from blivet3.size import Size
+    from blivet3.util import set_up_logging
+    BLIVET_PACKAGE = 'blivet3'
+except ImportError:
+    from blivet import Blivet
+    from blivet.callbacks import callbacks
+    from blivet.flags import flags as blivet_flags
+    from blivet.formats import get_format
+    from blivet.partitioning import do_partitioning
+    from blivet.size import Size
+    from blivet.util import set_up_logging
+    BLIVET_PACKAGE = 'blivet'
+
+from ansible.module_utils.basic import AnsibleModule
+
+blivet_flags.debug = True
+set_up_logging()
+log = logging.getLogger(BLIVET_PACKAGE + ".ansible")
+use_partitions = None  # create partitions on pool backing device disks?
+disklabel_type = None  # user-specified disklabel type
+
+
+class BlivetAnsibleError(Exception):
+    pass
+
+
+class BlivetVolume:
+    def __init__(self, blivet_obj, volume, bpool=None):
+        self._blivet = blivet_obj
+        self._volume = volume
+        self._blivet_pool = bpool
+        self._device = None
+
+    @property
+    def ultimately_present(self):
+        """ Should this volume be present when we are finished? """
+        return (self._volume['state'] == 'present' and
+                (self._blivet_pool is None or self._blivet_pool.ultimately_present))
+
+    def _type_check(self):  # pylint: disable=no-self-use
+        """ Is self._device of the correct type? """
+        return True
+
+    def _get_device_id(self):
+        """ Return an identifier by which to try looking the volume up. """
+        return self._volume['name']
+
+    def _look_up_device(self):
+        """ Try to look up this volume in blivet's device tree. """
+        device = self._blivet.devicetree.resolve_device(self._get_device_id())
+        if device is None:
+            return
+
+        self._device = device
+
+        # check that the type is correct, raising an exception if there is a name conflict
+        if not self._type_check():
+            self._device = None
+            return  # TODO: see if we can create this device w/ the specified name
+
+    def _get_format(self):
+        """ Return a blivet.formats.DeviceFormat instance for this volume. """
+        fmt = get_format(self._volume['fs_type'],
+                         mountpoint=self._volume.get('mount_point'),
+                         label=self._volume['fs_label'],
+                         options=self._volume['fs_create_options'])
+        if not fmt.supported or not fmt.formattable:
+            # FAIL: fs type tools are not available
+            raise BlivetAnsibleError("required tools for file system '%s' are missing" % self._volume['fs_type'])
```

Comment:
This is not a helpful message as it does not say what tools are those (package name or at least the name of a command or a  Python module). Also, the role should itself install anything that is needed, so we should never hit this point if the module is executed from the role.

Reply:
It turns out that the missing tool (I saw it actually happen) was `libblockdev-lvm`. This is even more confusing, as I would expect "tool" to refer to an executable, most likely those provided by the `lvm2` package, not to a library.

Reply:
When I install `libblockdev-lvm` manually, I get https://github.com/linux-system-roles/storage/pull/28/files#r310656215 , at which point I am lost, because I don't know how to obtain `libblockdev-dm`.

Reply:
fix in #33 - https://github.com/linux-system-roles/storage/pull/33/commits/3162351591c9d96c5f74c38c5c89adcdd796c4e7 . One should then not see the error message, still, is there aay to improve it, just in case, to point to exactly what needs to be installed?